head.S

```diff
@@ -164,9 +171,18 @@ GLOBAL(_entry)
 	push	$0
 	popf
 
+	cmpl	$0, boot_protocol(%ebp)
+	jne	1f
+
 	/* Jump to entry target - EBX: startup_32 ESI: ZP base EDX: LZ base */
 	mov	%ebp, %edx
 	jmp	*%ebx
+
+1:
+	mov	%ebx, %ecx
+	mov	%esi, %ebx
+	mov	$MULTIBOOT2_BOOTLOADER_MAGIC, %eax
+	jmp *%ecx
```

Comment:
`xchg %ebx, %esi` and `jmp *%esi` ?

It would also help to have a comment like above explaining the ABI in use.

Reply:
Done